ballerina/variable_validator.bal

```diff
@@ -154,4 +172,47 @@ class VariableValidator {
             self.errors.push(getErrorDetailRecord(message, location)); 
         }
     }
+
+    isolated function checkVariableUsageCompatibility(__InputValue[] input, parser:VariableDefinition varDef,
+                                                      parser:ArgumentNode argNode) {
+        __InputValue? inputValue = getInputValueFromArray(input, argNode.getName());
+        __Type? actualType = self.getTypeRecordFromTypeName(varDef.kind);
+        if inputValue is __InputValue && actualType is __Type {
+            if getOfType(inputValue.'type)?.name != getOfType(actualType)?.name {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            } else if inputValue.'type.kind == NON_NULL && actualType.kind != NON_NULL && varDef?.defaultValue is () {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            }
+        }
```

Comment:
What happens in the `else` part? Shouldn't we handle it?

- What if the `actualType` returns `()`?
- What if `inputValue` is `()`?

Reply:
Both cases are already handled by the `validator_visitor`.

Reply:
We use `variable_validator` before the `validator_visitor` right? Is that okay? I mean doesn't it cause any issues here?

Reply:
As we discussed offline, `else` logic is skipped here since it checks at `validator_visitor`.